facilitator/src/gcp_oauth.rs

```diff
@@ -393,13 +452,14 @@ impl GcpOauthTokenProvider {
     /// The returned value is an owned reference because the token owned by this
     /// struct could change while the caller is still holding the returned token
     fn ensure_default_account_token(&mut self) -> Result<String> {
-        if let Some(token) = &self.default_account_token {
+        if let Some(token) = &*self.default_account_token.read().unwrap() {
             if !token.expired() {
                 debug!(self.logger, "cached default account token is still valid");
                 return Ok(token.token.clone());
             }
         }
 
+        let mut default_account_token = self.default_account_token.write().unwrap();
         let http_response = self.default_token_provider.default_token()?;
```

Comment:
Now that we acquired the write lock, we need to double check that it's still expired. If we have contention, we could have multiple threads who all independently observed in the `read` lock that it was expired, but one already made it into the `write` and updated it.

At least, I'm pretty sure that's how the semantics of the above block work, since we don't keep the RwLockReadGuard in-scope.

Reply:
This is a very good point. The tricky thing is that I don't think you can acquire the write lock while the read lock is held. I think that at the time that I wrote this, I convinced myself that it's OK for two threads to race to acquire the write lock and update the access token, because regardless of who wins (1) both threads leave with a valid access token and (2) self.default_account_token will contain a valid access token. However I think the more obvious and safer thing to do would be to use an `std::sync::Mutex` here instead of `RwLock` so that we don't have to juggle two locks, at the cost of more lock contention on the read path.

Reply:
All you have to do is double-check `token.expired()` after acquiring the write lock and before proceeding. It is a little more expensive but it follows the general threading model of keeping the common case fast, and assuming success but checking for failure.

Changing to a Mutex seems overkill if this is the only place we encounter issues.

Reply:
I must be missing something. If `token` is in scope past the acquisition of `self.default_account_token.write()`, then `self.default_account_token().read()` will still be held and we'll deadlock because you can't acquire a write lock if there are any live readers.

Reply:
Insert after line 462 a check of expired and return early if the precondition no longer holds. It's a waste of time having obtained a write lock only to abandon ship, but it is consistent.

Reply:
Oh, duh, I was missing something: for some reason my brain decided that only the read lock can let you read but of course you can check the value with the write lock.